src/libcore/slice/mod.rs

```diff
@@ -1502,9 +1502,10 @@ unsafe impl<'a, T> TrustedLen for IterMut<'a, T> {}
 // Return the arithmetic difference if `T` is zero size.
 #[inline(always)]
 fn ptrdistance<T>(start: *const T, end: *const T) -> usize {
-    let diff = (end as usize).wrapping_sub(start as usize);
-    let size = mem::size_of::<T>();
-    diff / (if size == 0 { 1 } else { size })
+    match start.offset_to(end) {
+        Some(x) => x as usize,
+        None => end as usize - start as usize,
```

Comment:
For the ZST case the pointers are not assumed to be non-null, presumably because the wrapping arithmetic offset on their pointers can make them wrap around to zero. (There's also no NonZero around the pointer in the iterator).

But in the current formulation the ZST iteration always starts at 0x1. But the maximum ZST vec length should be usize::MAX! Which brings the end pointer to wrap around. I had to experiment to show.

```rust
fn main() {
    unsafe {
        let v = vec![(); !0];
        println!("{:?}", v.iter().len());
        let iter_repr: [usize; 2] = std::mem::transmute(v.iter());
        println!("{:?}", iter_repr);
    }
}
```

Output of the program:

```rust
18446744073709551615
[1, 0]
```

So for ZST it does wrap around.